skimage/util/_montage.py

```diff
@@ -0,0 +1,252 @@
+       ---> K                     +---+---+
+
+    where the '*' patch will be determined by the `fill` parameter.
+
+    Parameters
+    ----------
+    arr_in : (K, M, N[, C]) ndarray
+        3-/4-dimensional input array representing an ensemble of `K` images
+        of equal shape.
+    fill : float or array-like of floats or 'mean', optional
+        Value to fill the padding areas and/or the extra tiles in
+        the output array. Has to be `float` for single channel collections.
+        For multichannel collections has to be an array-like of shape of
+        number of channels. If `mean`, uses the mean value over all images.
+    rescale_intensity : bool, optional
+        Whether to rescale the intensity of each image to [0, 1].
+    grid_shape : tuple, optional
+        The desired grid shape for the montage (ntiles_row, ntiles_column).
+        The default aspect ratio is square.
+    padding_width : int, optional
+        The size of the spacing between the tiles and between the tiles and
+        the borders. If non-zero, makes the boundaries of individual images
+        easier to perceive.
+    multichannel : boolean, optional
+        If True, the last `arr_in` dimension is threated as a color channel,
+        otherwise as spatial.
+
+    Returns
+    -------
+    arr_out : (K*M+p, K*N+p[, C]) ndarray
+        Output array with input images glued together (including padding `p`).
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skimage.util.montage import montage
+    >>> arr_in = np.arange(3 * 2 * 2).reshape(3, 2, 2)
+    >>> arr_in  # doctest: +NORMALIZE_WHITESPACE
+    array([[[ 0,  1],
+            [ 2,  3]],
+           [[ 4,  5],
+            [ 6,  7]],
+           [[ 8,  9],
+            [10, 11]]])
+    >>> arr_out = montage(arr_in)
+    >>> arr_out.shape
+    (4, 4)
+    >>> arr_out
+    array([[ 0,  1,  4,  5],
+           [ 2,  3,  6,  7],
+           [ 8,  9,  5,  5],
+           [10, 11,  5,  5]])
+    >>> arr_in.mean()
+    5.5
+    >>> arr_out_nonsquare = montage(arr_in, grid_shape=(1, 3))
+    >>> arr_out_nonsquare
+    array([[ 0,  1,  4,  5,  8,  9],
+           [ 2,  3,  6,  7, 10, 11]])
+    >>> arr_out_nonsquare.shape
+    (2, 6)
+
+    """
+    if multichannel:
+        arr_in = np.asarray(arr_in)
+    else:
+        arr_in = np.asarray(arr_in)[..., np.newaxis]
+
+    if arr_in.ndim != 4:
+        raise ValueError('Input array has to be either 3- or 4-dimensional')
+
+    n_images, n_rows, n_cols, n_chan = arr_in.shape
+
+    if grid_shape:
+        ntiles_row, ntiles_col = map(int, grid_shape)
+    else:
+        ntiles_row = ntiles_col = int(np.ceil(np.sqrt(n_images)))
+
+    # Rescale intensity if necessary
+    if rescale_intensity:
+        for i in range(n_images):
+            arr_in[i] = exposure.rescale_intensity(arr_in[i])
+
+    # Calculate the fill value
+    if fill == 'mean':
+        fill = arr_in.mean(axis=(0, 1, 2))
+    fill *= np.ones(1)
```

Comment:
If `arr_in` is grayscale and `fill` is a scalar, it is neccessary to upcast `fill` to a ndarray of ndim 1 (np.asarray(scalar) returns an array of ndim 0 and doesn't allow indexing). The indexing is used in https://github.com/scikit-image/scikit-image/pull/2626/files/95a8513cfaee0e99cf85f75c091a7575e89499c8#diff-81a8a5406b37c24a621cf34d9208fcc0R115